### PR TITLE
[DNM][WIP] mgr/dashboard_v2: Integrate dashboard_v2 into CMake

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Integrate dashboard_v2 into `ctest`. How to use run the test:
```
make mgr-dashboard_v2
ctest -V -R dashboard_v2
```

---
Also: 
* Add CMake target to build the frontend.
* Added new CMake flag `WITH_MGR_DASHBOARD_V2_FRONTEND`:
  Build the mgr/dashboard_v2 frontend using `npm install && npm run build`
* Set this flag to `OFF` when building packages.
* Removed creation of the frontend from `vstart.sh`

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

### CODE REVIEW

**Review by LenzGr — DISMISSED** [review]

File: src/pybind/mgr/dashboard_v2/run-tox.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+
+# run from the ceph-disk directory or from its parent
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/run-tox.sh]: ceph-disk?

File: src/pybind/mgr/dashboard_v2/tests/helper.py [hunk 2]
```diff
@@ -23,6 +24,8 @@ class ControllerTestCase(unittest.TestCase):
     DASHBOARD_HOST = os.environ.get('DASHBOARD_V2_HOST', "localhost")
     DASHBOARD_PORT = os.environ.get('DASHBOARD_V2_PORT', 8080)
 
+    pytestmark = pytest.mark.skip
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/tests/helper.py]: why do we need this?

Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/tests/helper.py]: To make the tests pass when calling make check. Should be removed in theutology. This is like a global off-switch for all tests that inherent from `ControllerTestCase`

File: src/pybind/mgr/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+add_subdirectory(dashboard_v2)
```
Comment (rjfd) [inline src/pybind/mgr/CMakeLists.txt:1]: add line break

File: ceph.spec.in [hunk 4]
```diff
@@ -895,6 +895,7 @@ cmake .. \
     -DWITH_EMBEDDED=OFF \
     -DWITH_MANPAGE=ON \
     -DWITH_PYTHON3=ON \
+    -DWITH_MGR_DASHBOARD_V2_FRONTEND=OFF \
```
Comment (LenzGr) [inline ceph.spec.in:898]: Do we need to add something similar to the DEB builds?

File: src/pybind/mgr/CMakeLists.txt (hunk 3 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/CMakeLists.txt:1]: fixed

File: src/pybind/mgr/dashboard_v2/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  ${CMAKE_SOURCE_DIR}/src/tools/setup-virtualenv.sh ${MGR_DASHBOARD_V2_VIRTUALENV} &&
+  ${MGR_DASHBOARD_V2_VIRTUALENV}/bin/pip install --no-index --use-wheel --find-links=file:${CMAKE_SOURCE_DIR}/src/pybind/mgr/dashboard_v2/wheelhouse -r requirements.txt
+  WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}/src/pybind/mgr/dashboard_v2
+  COMMENT "mgr-dashboard_v2 is being created")
+add_dependencies(tests mgr-dashboard_v2)
+
+include(Distutils)
+distutils_install_module(mgr_dashboard_v2)
+
+option(WITH_MGR_DASHBOARD_V2_FRONTEND "Build the mgr/dashboard_v2 frontend using `npm install && npm run build`" ON)
```
Comment (LenzGr) [inline src/pybind/mgr/dashboard_v2/CMakeLists.txt]: I would shorten that description to "Build the mgr/dashboard_v2 web frontend using `npm`"

**Comment by smithfarm** [comment]
The -D...=OFF thing needs to be added to debian/rules (extraopts) as well

File: ceph.spec.in (hunk 4 above)
Comment (rjfd) [reply ceph.spec.in:898]: Yes

Comment (sebastian-philipp) [reply ceph.spec.in:898]: fixed

File: src/pybind/mgr/dashboard_v2/CMakeLists.txt (hunk 5 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/CMakeLists.txt]: What about "Build the mgr/dashboard_v2 web frontend using npm. Requires internet access."?

**Comment by sebastian-philipp** [comment]
@smithfarm: fixed

**Comment by smithfarm** [comment]
@sebastian-philipp This is a preliminary step before adding the test to "make check", right? 

In other words, this change does not do `add_ceph_test(...)` so the test will not be run by "make check", correct?

Also, the commit message mentions "make test" - I'm not familiar with it, what does it do?

**Comment by smithfarm** [comment]
(I guess "make test" just builds the tests, but doesn't run them?)

**Comment by smithfarm** [comment]
Duh, should have read the code first: it's `make tests` to build the tests and `ctest` to run them.

@sebastian-philipp So maybe change "make test" to "make tests" in the commit message? Not sure.

**Comment by sebastian-philipp** [comment]
@smithfarm Hm. I did it like all other tox based tests. Does that mean, that these other tests are not executed as well?

**Comment by smithfarm** [comment]
@sebastian-philipp I wish I knew the answer. Does the test run when you do `./run-make-check.sh` in this branch?

**Review by smithfarm — APPROVED** [review]

**Review by LenzGr — DISMISSED** [review]

File: src/pybind/mgr/dashboard_v2/CMakeLists.txt [hunk 6]
```diff
@@ -10,3 +10,44 @@ add_dependencies(tests mgr-dashboard_v2)
...
+  frontend/src/*/*/*.html
+  frontend/src/*/*/*/*.ts
+  frontend/src/*/*/*/*.html
+  frontend/src/*/*/*/*/*.ts
+  frontend/src/*/*/*/*/*.html
+  frontend/src/*/*/*/*/*/*.ts
+  frontend/src/*/*/*/*/*/*.html)
+
+if(NOT CMAKE_BUILD_TYPE STREQUAL Debug)
+  set(npm_command npm install && npm build -- --prod)
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/CMakeLists.txt:36]: the correct frontend build command is: `npm run build -- --prod`

File: src/pybind/mgr/dashboard_v2/CMakeLists.txt [hunk 7]
```diff
@@ -10,3 +10,44 @@ add_dependencies(tests mgr-dashboard_v2)
...
+else()
+  set(npm_command npm install && npm build)
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/CMakeLists.txt:38]: the correct frontend build command is: `npm run build`

**Comment by sebastian-philipp** [comment]
Was cherry-picked instead. close.

**Pull request closed without merging**